### PR TITLE
676 - translations!

### PR DESCRIPTION
**Pull request by bent0b0x**
**Issue**
#676 

**Translation System**
- have `express` server inject locale into page, via `window.paratiiPortal.locale`
- create translation files in the form of `.json` files. Keys can be nested as deep as we need and can take dynamic options to be injected uniquely into the strings for each language
- Create react component that takes in a translation key and renders the correct text

**Demo**

image

You can't tell, but the circled text comes from this new system 😄 

**TODO**
- [x] inject translations for specified locale from the server so that we do not include every language in the bundle
- [x] move all english text to translation file! 👈 will do in separate PRs
- [x] figure out why translation changes don't live-reload in development mode  (not required, but is a nice-to-have) 👈 the translations are injected by the server now; we will not get live reloading w/ them which is fine
- [x] fall back to english (or whatever we want the default to be) if we are missing a translation in a particular language 👈 deep merge the files on the server before injecting into page? 3d7ccb8
- [x] precompile translations so they don't have to be merged and stringified at request-time 👈 added cacheing/memoization
- [x] figure out way to inject components into translations (will be covered by https://github.com/Paratii-Video/paratii-portal/issues/689) 👈 completed by 5167d1a

cc @jellegerbrandy let me know what you think of this system. Can definitely change it as needed.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -148,20 +148,22 @@
...
     "lodash.debounce": "^4.0.8",
+    "node-polyglot": "^2.2.2",
     "nodemailer": "^4.6.4",
     "paratii-js": "github:Paratii-Video/paratii-js#a6e9ea8c24768f86650b18d92d2946d988fd860b",
     "paratii-mediaplayer": "3.1.2",
     "player.js": "^0.1.0",
     "promise-polyfill": "^7.1.2",
     "query-string": "^6.0.0",
-    "react": "16.2.0",
+    "react": "16.3",
```
Comment (bent0b0x) [inline package.json]: This upgrade is needed to take advantage of the new [context API](https://reactjs.org/docs/context.html), which we take advantage of to inject the translator into components no matter how deeply nested they are

File: src/scripts/components/pages/NotFound.js [hunk 2]
```diff
@@ -30,7 +34,9 @@ class NotFound extends Component<Props, void> {
   render () {
     return (
       <Wrapper>
-        <Title purple>Oooooops, page not found</Title>
+        <Title purple>
+          <TranslatedText message="notFound.title" />
```
Comment (bent0b0x) [inline src/scripts/components/pages/NotFound.js]: First example usage of the basic `<TranslatedText /> ` component

File: src/scripts/components/translations/ConnectToTranslator.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+const ConnectToTranslator = Component => {
+  const WrappedComponent = props => (
+    <TranslationContext.Consumer>
+      {translator => <Component {...props} translator={translator} />}
+    </TranslationContext.Consumer>
+  )
+
+  return WrappedComponent
+}
+export default ConnectToTranslator
```
Comment (bent0b0x) [inline src/scripts/components/translations/ConnectToTranslator.js]: Higher order component that connects a component to the translator specified in the translation context

File: src/scripts/utils/TranslationUtils.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+const DEFAULT_LOCALE = SUPPORTED_LOCALES.EN
+
+const getTranslations = ({ locale }): Object => {
+  switch (locale) {
+    case SUPPORTED_LOCALES.EN:
+    default:
+      return require('translations/en.json')
+  }
```
Comment (bent0b0x) [inline src/scripts/utils/TranslationUtils.js]: I am going to move this to the server

File: src/server/app.js [hunk 5]
```diff
@@ -31,6 +33,7 @@ app.engine(
 )
 
 app.use(compression())
+app.use(createLocaleMiddleware())
```
Comment (bent0b0x) [inline src/server/app.js]: This middleware grabs the locale from request headers

File: src/scripts/index.js [hunk 6]
```diff
@@ -8,29 +8,39 @@ import 'promise-polyfill/src/polyfill'
...
-        </AppContainer>
-      </Provider>,
+      <TranslationContext.Provider value={translator}>
+        <Provider store={store}>
+          <AppContainer>
+            <NextRoot />
+          </AppContainer>
+        </Provider>
+      </TranslationContext.Provider>,
       getRoot()
```
Comment (bent0b0x) [inline src/scripts/index.js]: I also plan on de-duping this component hierarchy on this branch

**Comment by jellegerbrandy** [comment]
looks good. As long as it is using i18n I'm happy :-)

File: src/server/utils/getTranslations.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  switch (locale) {
+    case SUPPORTED_LOCALES.EN:
+    default:
+      localeTranslations = null
+      break
+  }
+
+  const finalTranslations = localeTranslations
+    ? mergeDeep(defaultTranslations, localeTranslations)
+    : defaultTranslations
```
Comment (bent0b0x) [inline src/server/utils/getTranslations.js]: This uses english translations as the default. I think ideally these merges would be pre-compiled. I'll take care of this before merging

**Comment by bent0b0x** [comment]
@jellegerbrandy I think this is good to merge whenever; not all of the translations are complete, obviously, but this sets the system up and we can add the rest of the translations (soon) after.

File: src/scripts/components/pages/SearchResults.js [hunk 8]
```diff
@@ -99,8 +100,12 @@ class SearchResults extends React.Component<Props, void> {
 
     return (
       <SearchTerm>
-        <SearchTermPrompt>Results for: </SearchTermPrompt>
-        {searchTerm}
+        <SearchTermPrompt>
+          <TranslatedText
+            message="search.results.resultsFor"
+            options={{ term: searchTerm }}
```
Comment (bent0b0x) [inline src/scripts/components/pages/SearchResults.js]: We can pass variables in

File: src/scripts/utils/translations/RawTranslatedText.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import React from 'react'
+import { renderToString } from 'react-dom/server'
+
+import TranslatedText from 'components/translations/TranslatedText'
+
+const RawTranslatedText = props => renderToString(<TranslatedText {...props} />)
+
+export default RawTranslatedText
```
Comment (bent0b0x) [inline src/scripts/utils/translations/RawTranslatedText.js]: This is needed when we have to pass translations to html element attributes, e.g. `placeholder` for an `input`

File: src/server/middleware/TranslationsMiddleware.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    : memoize(getTranslations)
+
+module.exports = () => (req, res, next) => {
+  const locale = req.locale.language
+
+  const translations = getTranslationsForLocale(locale)
+
+  req.translations = translations
+
+  next()
```
Comment (bent0b0x) [inline src/server/middleware/TranslationsMiddleware.js]: This attaches the translations json blob to every request on the server

File: src/translations/en.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    "title": "Oooooops, page not found",
+    "description": "This page is not available yet. A passionate team is either building it, or unaware of it, as of now. Want to help us move faster? Please report the issue on our Github:",
+    "linkText": "Paratii-Video"
+  },
+  "search": {
+    "input": {
+      "placeholder": "Search"
+    },
+    "results": {
+      "resultsFor": "Results for: %{term}",
```
Comment (bent0b0x) [inline src/translations/en.json]: variables 💯

File: webpack-utils/TranslationsWebpackPlugin.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+const defaultTranslations = require('../src/translations/en.json')
+
+module.exports = () =>
+  new CopyWebpackPlugin([
+    {
+      from: './src/translations/**/*.json',
+      to: './translations/[name].json',
+      transform: content =>
+        JSON.stringify(mergeDeep(defaultTranslations, JSON.parse(content)))
```
Comment (bent0b0x) [inline webpack-utils/TranslationsWebpackPlugin.js]: Here we merge translation files such that we default to english for any locale missing a particular translation value

File: src/translations/en.json [hunk 13]
```diff
@@ -0,0 +1,33 @@
+{
+  "footer": {
+    "discover_html": "Discover all about Paratii on %{homePageLink}",
+    "homePage": "paratii.video",
+    "betaTool_html": "v0.0.1: This is a beta tool! We &hearts; to get feedback on Telegram %{telegramBrazilLink} or %{telegramEnglishLink}",
```
Comment (bent0b0x) [inline src/translations/en.json]: I have implemented a solution for #689. If we add an affix of `_html` to our keys, then the `<TranslatedText />` component will convert all passed in options, even if they contain JSX, to raw strings and inject them (unsafely) as html. 

We cannot use `_html` translations with anything that contains user input.

File: src/scripts/components/structures/footer/MainFooter.js [hunk 14]
```diff
@@ -48,30 +50,48 @@ class MainFooter extends Component<Props, void> {
...
+                  anchor
+                  href="https://t.me/joinchat/EWZMBQ9mnqJ1--NKHpyF8A"
+                  target="_blank"
+                  underline
+                >
+                  <TranslatedText message="footer.english" />
+                </Anchor>
+              )
+            }}
+          />
```
Comment (bent0b0x) [inline src/scripts/components/structures/footer/MainFooter.js]: Example of how we can translate something and include JSX/react in the options

File: src/scripts/utils/translations/renderRawTranslatable.js [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+import React from 'react'
+import { renderToString } from 'react-dom/server'
+import { ThemeProvider } from 'styled-components'
+
+import { paratiiTheme } from 'constants/ApplicationConstants'
+
+const renderRawTranslatable = (children: any) =>
+  renderToString(<ThemeProvider theme={paratiiTheme}>{children}</ThemeProvider>)
+
+export default renderRawTranslatable
```
Comment (bent0b0x) [inline src/scripts/utils/translations/renderRawTranslatable.js]: This is a util that is used to inject html into translations.

**Comment by bent0b0x** [comment]
@jellegerbrandy I think this is good to merge, but the tests are failing because, I believe, the uploader is not working. 

This PR now includes a solution for rendering html/jsx in translations, as described in #689

**Pull request merged**